### PR TITLE
refactor(initramfs): Compose Install better

### PR DESCRIPTION
**Pull request by bradbeam**
Prep work for #623 

Signed-off-by: Brad Beam <brad.beam@talos-systems.com>

### CODE REVIEW

**Comment by bradbeam** [comment]
Still needs to be tested/etc but wanted to get some eyes on it.

The thought is if we can make the Install a method off target, we can reuse the work done in Prepare() to generate the install manifest/targets. That should allow us to consume this for the upgrade functionality ( we'd need to mimic the work that is done in the `Install(user data)` func ).

File: internal/pkg/install/prepare.go [hunk 1]
```diff
@@ -410,3 +416,103 @@ func (t *Target) Format() error {
...
+		case "https":
+			log.Printf("downloading %s\n", artifact)
+
+			sourceFile, err = download(u, t.MountPoint, "")
+			if err != nil {
+				return err
+			}
+
+			// TODO why dont we use defer?
+			// maybe for single file downloads ( like kernel )
```
Comment (bradbeam) [inline internal/pkg/install/prepare.go]: I might add in some work to try to clean up all the cleanup conditionals below.

File: internal/pkg/install/prepare.go [hunk 2]
```diff
@@ -410,3 +437,108 @@ func (t *Target) Format() error {
...
+			log.Printf("downloading %s\n", artifact)
+
+			sourceFile, err = download(u, t.MountPoint)
+			if err != nil {
+				return err
+			}
+
+			// TODO why dont we use defer?
+			// maybe for single file downloads ( like kernel )
+			// defer os.Remove(sourceFile.Name())
```
Comment (andrewrynhard) [inline internal/pkg/install/prepare.go]: can we remove the commented code?

File: internal/pkg/install/prepare_test.go [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+	"path/filepath"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/suite"
+	"github.com/talos-systems/talos/pkg/userdata"
+	"gopkg.in/yaml.v2"
+)
+
+// NewManifest
```
Comment (andrewrynhard) [inline internal/pkg/install/prepare_test.go]: nit: remove

File: internal/pkg/install/prepare_test.go [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+// target.Install()
```
Comment (andrewrynhard) [inline internal/pkg/install/prepare_test.go]: nit: remove

File: internal/pkg/install/prepare_test.go [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+
+// verifyXXDevice
```
Comment (andrewrynhard) [inline internal/pkg/install/prepare_test.go]: nit: remove

File: internal/pkg/install/prepare_test.go [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+	suite.Suite
+}
+
+func TestValidateSuite(t *testing.T) {
+	suite.Run(t, new(validateSuite))
+}
+
+func (suite *validateSuite) TestNewManifest() {
+	// suite.Require().NoError(err)
+	// assert.Equal(suite.T(), "test", kubeadm.CertificateKey)
```
Comment (andrewrynhard) [inline internal/pkg/install/prepare_test.go]: nit: remove these

File: internal/pkg/install/prepare_test.go [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+	// hard fail here
+	data.Install.Root = nil
+	suite.Require().Error(VerifyRootDevice(data))
+}
+
+func (suite *validateSuite) TestTargetInstall() {
+	/*
+	   func TempDir(dir, prefix string) (name string, err error)
+	   func TempFile(dir, pattern string) (f *os.File, err error)
+	*/
```
Comment (andrewrynhard) [inline internal/pkg/install/prepare_test.go]: nit: remove

File: internal/pkg/install/install.go [hunk 8]
```diff
@@ -41,108 +40,28 @@ func Install(args string, data *userdata.UserData) (err error) {
...
-				log.Printf("copying %s to %s\n", artifact, filepath.Join(dest, filepath.Base(source)))
-				sourceFile, err = os.Open(source)
-				if err != nil {
-					return err
-				}
-
-				destFile, err = os.Create(filepath.Join(dest, filepath.Base(source)))
-				if err != nil {
+	manifest := NewManifest(data)
+	// manifest.Targets = map[string(device)][]*Target
```
Comment (andrewrynhard) [inline internal/pkg/install/install.go]: nit: remove

**Comment by andrewrynhard** [comment]
LGTM! Rebase and merge whenever you can.

**Pull request merged**